packages/network/src/connection/Connection.ts

```diff
@@ -201,8 +215,43 @@ export class Connection extends ConnectionEmitter {
         })
     }
 
+    createDeferredConnectionAttempt(): void {
+        this.deferredConnectionAttempt = new DeferredConnectionAttempt(this.peerInfo.peerId)
+    }
+
+    setDeferredConnectionAttempt(attempt: DeferredConnectionAttempt | null): void {
+        this.deferredConnectionAttempt = attempt
+    }
+
+    getDeferredConnectionAttempt(): DeferredConnectionAttempt | null {
+        return this.deferredConnectionAttempt
+    }
+
+    stealDeferredConnectionAttempt(): DeferredConnectionAttempt | null {
+        const att = this.deferredConnectionAttempt
+        this.deferredConnectionAttempt = null
+        return att
+    }
+
+    private emitOpen() {
+        if (this.deferredConnectionAttempt) {
+            const def = this.deferredConnectionAttempt
+            this.deferredConnectionAttempt = null
+            def.emit('open')
+        }
+        this.emit('open')
+    }
+
+    private emitClose(reason: any) {
+        if (this.deferredConnectionAttempt) {
+            const def = this.deferredConnectionAttempt
+            this.deferredConnectionAttempt = null
+            def.emit('close', reason)
+        }
```

Comment:
I would disagree with this change. "steal" in the name indicates a destructive operation, and I think we should first test the precondition, and only when all is ok, do the destructive operation. (even though the proposed change works, it looks dangerous to whoever is reading the code)

Reply:
I re-structured the DefrredConnectionAttempt class to make it more pretty. I made the event emitter a private field instead of deriving from it, and created new public methods for accessing: resolve() and reject() to make the promise analogy more clear.